pcdsdevices/interface.py

```diff
@@ -0,0 +1,893 @@
+"""
+Module for defining bell-and-whistles movement features
+"""
+import time
+import fcntl
+import logging
+import numbers
+import signal
+import re
+from contextlib import contextmanager
+from pathlib import Path
+from threading import Thread, Event
+from types import SimpleNamespace, MethodType
+from weakref import WeakSet
+
+import yaml
+from bluesky.utils import ProgressBar
+from ophyd.device import Kind
+from ophyd.status import wait as status_wait
+
+from . import utils as util
+
+logger = logging.getLogger(__name__)
+engineering_mode = True
+
+OphydObject_whitelist = ["name", "connected", "check_value", "log"]
+BlueskyInterface_whitelist = ["trigger", "read", "describe", "stage",
+                              "unstage"]
+Device_whitelist = ["read_attrs", "configuration_attrs", "summary",
+                    "wait_for_connection", "stop", "get", "configure"]
+Signal_whitelist = ["value", "put"]
+Positioner_whitelist = ["settle_time", "timeout", "egu", "limits", "move",
+                        "position", "moving"]
+
+
+class BaseInterface:
+    """
+    Interface layer to attach to any Device for SLAC features.
+
+    This class defines an API and some defaults for filtering tab-completion
+    results for new users to avoid confusion. The API involves setting the
+    tab_whitelist attribute on any subclass of BaseInterface. When in
+    non-engineering mode, only elements on the whitelists will be displayed to
+    the user.
+
+    Attributes
+    ----------
+    tab_whitelist: list
+        list of string regex to show in autocomplete for non-engineering mode
+    """
+    tab_whitelist = (OphydObject_whitelist + BlueskyInterface_whitelist +
+                     Device_whitelist + Signal_whitelist +
+                     Positioner_whitelist)
+    _filtered_dir_cache = None
+
```

Comment:
I haven't given this as much thought as you, but should maybe `tab_component_names` be `True` by default?

Reply:
Maybe. I was seeing long lists of "omitted" components so I defaulted to `False`, but then later I switched the `tab_component_names` to ignore `omitted` components without revisiting the default.

Reply:
The other thing to consider is that, with the nature of whitelisting, the package is more consistent if showing anything on tab defaults to `False`.

Reply:
True!